### PR TITLE
gh-118331: Don't raise an error if tuple allocation fails when clearing weakrefs

### PR DESCRIPTION
**Pull request by mpage**
It's not safe to raise an exception in `PyObject_ClearWeakRefs()` if one is not already set. `PyObject_ClearWeakRefs()` may be called (transitively) by `_Py_Dealloc()`, which contains https://github.com/python/cpython/blob/5a90de0d4cbc151a6deea36a27eb81b192410e56/Objects/object.c#L2843-L2860

Additionally, make sure we clear the weakrefs when tuple allocation fails.

This bug predates gh-111926. It [exists in 3.12](https://github.com/python/cpython/blob/186fbe82f7b3124835bcd3cfc4a03139ea8ba034/Objects/weakrefobject.c#L977-L981) and likely in earlier versions too. If it's getting tickled now, I suspect it's because we always allocate a tuple now when clearing weakrefs that have callbacks, whereas previously we only did so if there was more than one weakref with a callback, making it more likely that we fail to allocate a tuple when running `test_repl.TestInteractiveInterpreter.test_no_memory`.

* Issue: gh-118331

### CODE REVIEW

**Review by Fidget-Spinner — APPROVED** [review]

**Comment by hugovk** [comment]
Thanks, this fixes gh-118331 for me.

**Review by colesbury — APPROVED** [review]
Thanks!

File: Objects/weakrefobject.c [hunk 1]
```diff
@@ -1016,7 +1016,13 @@ PyObject_ClearWeakRefs(PyObject *object)
...
     PyObject *tuple = PyTuple_New(num_weakrefs * 2);
     if (tuple == NULL) {
-        _PyErr_ChainExceptions1(exc);
+        _PyWeakref_ClearWeakRefsExceptCallbacks(object);
+        if (exc == NULL) {
+            PyErr_WriteUnraisable(NULL);
+        }
+        else {
+            _PyErr_ChainExceptions1(exc);
+        }
```
Comment (colesbury) [inline Objects/weakrefobject.c]: Hmm... I think this should instead be:

```suggestion
        PyErr_WriteUnraisable(NULL);
        PyErr_SetRaisedException(exc);
```

In other words, write the unraisable exception and then restore the previous active exception, if any.

I don't think we should chain exceptions. The assertion in `_Py_Dealloc` checks that the exception did not change, but as written we would end up changing the exception if there was previously an active one.

Comment (colesbury) [reply Objects/weakrefobject.c]: For an example of this pattern see _PyGen_Finalize:

https://github.com/python/cpython/blob/23d0371bb99b1df183c36883e256f82fdf6a4bea/Objects/genobject.c#L81-L89

**Review by colesbury — CHANGES_REQUESTED** [review]

**Comment by mpage** [comment]
I have made the requested changes; please review again.

**Review by colesbury — APPROVED** [review]

**Pull request merged**